### ISSUE TITLE
Annotations for incompatible native histogram operations with binary operators have confusing message

### ISSUE THREAD

**Issue by charleskorn** [prometheus/prometheus#14669]
### What did you do?

I performed the following queries:

`series_with_exponential_native_histograms + series_with_custom_buckets_native_histograms`

`series_with_custom_buckets_native_histograms_layout_1 + series_with_custom_buckets_native_histograms_layout_2`

As the names suggest, these series contained native histograms with exponential and custom buckets.

### What did you expect to see?

Warnings that either include the appropriate metric names, or don't mention metric names at all.

For example, the first case could return a warning like `PromQL warning: vector contains a mix of histograms with exponential and custom buckets schemas for binary operation (1:1)`.

### What did you see instead? Under which circumstances?

The first query result included the warning `PromQL warning: vector contains a mix of histograms with exponential and custom buckets schemas for metric name "" (1:1)`.

The second query result included the warning `PromQL warning: vector contains histograms with incompatible custom buckets for metric name "" (1:1)`.

**Comment by beorn7** [comment]
Clearly, the intention was to have the metric field filled.

@zenador would you like to investigate?

**Comment by charleskorn** [comment]
> Clearly, the intention was to have the metric field filled.

Which metric name would we use for binary operations though? The left side? The right? Both?(What happens if the left or right side don't have metric names?)

**Comment by beorn7** [comment]
Ideally, we would have both names, or even be more specific about the exact label match that leads to the collision. Or maybe that's way too much noise and we should be more general. That needs to be explored.

In any case, it's not just about the name. The wording of the annotation ("vector contains a mix…") is just not right for bin-ops, where the mismatch is between elements from _different_ vectors.

**Comment by NeerajGartia21** [comment]
Hi @charleskorn , @beorn7 . I investigated on this and I think [this](https://github.com/prometheus/prometheus/blob/d4994e5bc44490441a6a6ac05331cc6fbabae0f5/promql/engine.go#L3263) function handles all the error related to binary operations between vectors. Here deliberately the metric name is set to empty string, so that's the reason for empty metric name in the annotation.

**Comment by NeerajGartia21** [comment]
Yeah, since these annotations are added in Vector binary operations error handler, we can introduce a new annotation for this case, like "binary operations is performed on histograms with exponential and custom buckets schemas (between {LHS expr} and {RHS expr})". 
What are your thoghts on this?

**Comment by beorn7** [comment]
@NeerajGartia21 since you added annotation for other bin ops, I'll assign this to you for now.

**Comment by NeerajGartia21** [comment]
Thanks @beorn7 for assigning it to me.  I'll have a look over it.

**Comment by NeerajGartia21** [comment]
For this Issue, I found three options.
- We can add the metric name to the current annotation like- `PromQL warning: vector contains a mix of histograms with exponential and custom buckets schemas for metric name "metricName" (1:1)`. But It is not well suited for binary operations and does not convey the message well.
- We can use the existing info annotation like- `incompatible sample types encountered for binary operator: exponential bucket histogram $OP custom bucket histogram`. It seems good but it is an info annotation and currently we are using warn annotation.
- Thirdly we can create another warn annotation like `$BINOP contains a mix of histogram with custom bucket and exponential bucket for the metric : $METRIC_NAME`

What are your opinions on these?

**Comment by beorn7** [comment]
> We can use the existing info annotation like- incompatible sample types encountered for binary operator: exponential bucket histogram $OP custom bucket histogram. It seems good but it is an info annotation and currently we are using warn annotation.

If possible, I think we should use the same approach as for this bin-op info annotation, just on warn level and maybe change the text from "incompatible sample type" to "incompatible bucket layout". Also, if we have two custom bucket histograms, it wouldn't be very understandable to say "custom bucket histogram $OP custom bucket histogram". It had to be something like "incompatible custom bucket layout encountered for binary operator $OP". Or keep it simple and just say "incompatible custom bucket layout encountered for binary operator $OP" without explaining the arguments. I assume this will be rare enough that people should be fine with it.

**Comment by NeerajGartia21** [comment]
Yeah, I think we can go with `incompatible bucket layout encountered for binary operator $OP for metric : $METRICNAME`. It will be sufficient enough to convey the message ig.
